### PR TITLE
feat: alternate path longer than the location finder

### PR DESCRIPTION
**Pull request by BonyBabu**

### CODE REVIEW

File: encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperty.java [hunk 1]
```diff
@@ -81,7 +81,10 @@
     PHYSICAL_FORMAT_VERSION("PhysicalFormatVersion", PropertyType.INTEGER_BY_ID, -1),
 
     /** The COM p_ tim e_4_ cache. */
-    COMP_TIME_4_CACHE("CompTime4Cache", PropertyType.INTEGER, 0);
+    COMP_TIME_4_CACHE("CompTime4Cache", PropertyType.INTEGER, 0),
+
+    ALTERNATE_PATH_RELATIVE_LOWER_THRESHOLD("AlternatePathRelativeLowerThreshold",PropertyType.INTEGER,-1);
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperty.java]: 3760b9e915acdf79c7fb28b3b3b719238f518de8

File: encoder/src/main/java/openlr/encoder/routesearch/RouteSearch.java [hunk 2]
```diff
@@ -134,7 +131,7 @@ public RouteSearch(final List<? extends Line> loc)
         endLoopIndex = checkLoopAtEnd();
     }
 
-    /**
+    /**ß
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/routesearch/RouteSearch.java]: I think you inserted a character here by accident.

Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/RouteSearch.java]: d2e779368ba0835d17385a622cc639523649e302

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    public static SecondShortestRouteChecker on(List<? extends Line> location, double relativeThreshold) throws OpenLREncoderProcessingException {
+        if (location.isEmpty()) {
+            throw new OpenLREncoderProcessingException(OpenLREncoderProcessingException.EncoderProcessingError.NO_ROUTE_FOUND_ERROR);
+        }
+        GeoCoordinates destinationStart = location.get(location.size() - 1).getStartNode().getGeoCoordinates();
+        int locationLength = location.stream().mapToInt(Line::getLineLength).sum();
+
+        int maxLengthAllowed = locationLength + (int) (locationLength * relativeThreshold);
+
+        BiFunction<Integer, Integer, Boolean> verifyThreshold = (Integer lengthAlongSecondShortestRoute, Integer lengthAlongLocation) -> {
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: You could just make this function an instance method and make the `relativeThreshold` a member variable.  I think defining it as a `BiFunction` member variable is a bit unusual and requires a bit more mental parsing :-)

```
private boolean verifyThreshold(int lengthAlongSecondShortestRoute, int lengthAlongLocation) {
    return lengthAlongSecondShortestRoute < (lengthAlongLocation + (int) (lengthAlongLocation * this.relativeThreshold);
}
```

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        while (to.getPrevious() != null && to.getPrevious().getLine().getID() != deviationStart) {
+            secondShortestRouteLength += to.getPrevious().getLine().getLineLength();
+            to = to.getPrevious();
+        }
+        return secondShortestRouteLength;
+    }
+
+    /**
+     *
+     * @param closedSet set of road segments which
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: `set of road segments which......`

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+     * @param routeSearchData priority queue
+     * @param index index of the road segment along the location where the
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: `index of road segment along the location where the......`

File: encoder/src/test/java/openlr/encoder/postprocessor/AlternatePathLrpProcessorTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        this.map = OpenLRMapDatabaseAdaptor.from(mapFile);
+    }
+
+    @Test
+    public void verifyAlternatePathHandlerProcess() throws OpenLRProcessingException
+    {
+        Configuration encoderConfig = OpenLRPropertiesReader.loadPropertiesFromFile(new File(TestMapStubTest.class.getClassLoader().getResource("OpenLR-Encoder-Properties.xml").getFile()));
+        encoderConfig.setProperty("LrpAlternativePath",true);
+        encoderConfig.setProperty("AlternatePathRelativeThreshold",.20);
+        OpenLREncoderProperties properties = new OpenLREncoderProperties(encoderConfig, new ArrayList< PhysicalEncoder >());
```
Comment (stephencurran) [inline encoder/src/test/java/openlr/encoder/postprocessor/AlternatePathLrpProcessorTest.java]: You don't need the explicit type `PhysicalEncoder` you can leave empty.

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <h1>Processor to insert intermediate points along the shortest route between adjacent location reference points
+ * where there are alternate paths with length under the threshold exist</h1>
+ */
+public class AlternatePathLrpProcessor implements LrpProcessor {
+    private OpenLREncoderProperties properties;
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: These properties are used to create LRPs

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                List<Line> connectedRoute = new ArrayList<>();
+                for (Line line : oldRoute) {
+                    connectedRoute.add(line);
+                }
+                connectedRoute.add(firstLineOfNextLrp);
+                List<Integer> intermediateLrpPositions = determineNewIntermediatePoints(connectedRoute);
+                if (intermediateLrpPositions.isEmpty()) {
+                    checkedList.add(lrp);
+                } else {
+                    checkedList.addAll(createNewLRPs(lrp.getRoute(), intermediateLrpPositions));
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: So in case we find intermediates we need to place new LRPs in the place of already existing LRPs because the route that this LRP represent will cnahge

File: encoder/src/main/java/openlr/encoder/routesearch/IntermediateHandler.java [hunk 9]
```diff
@@ -130,17 +141,17 @@ public final RouteSearchResult checkIntermediate(
...
+                // the new line is a direct successor of the last element found
+                // so the new line is now the last element found, we can
+                // continue calculating the route
+                lastElemPos++;
+                lastElemOnLocation = actualElement;
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("next line in sub route found: "
+                            + actualElement.getLine().getID());
+                }
+            } else {
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/IntermediateHandler.java]: That seems to be the default case

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+     * Generate a linked list of location reference points where No Partial or fully joined  alternate path of length under the given threshold
+     * exist between the road segments of adjacent location reference points
+     * @param lrps linked list of estimated location reference points
+     * @return linked list of revised lrp point linked list with intermediate points
+     * @throws OpenLRProcessingException
+     */
+    public List<LocRefPoint> process(List<LocRefPoint> lrps) throws OpenLRProcessingException {
+
+        if (properties.insertLrpAtAlternatePath()) {
+            List<LocRefPoint> checkedList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * @param route route between the adjacent location reference points(including the road segment of the destination lrp)
+     * @param lrpPositions indices of the road segment where to insert the intermediate points.
+     * @return list of location reference points with intermediate points
+     * @throws OpenLRProcessingException
+     */
+    private List<LocRefPoint> createNewLRPs(List<Line> route, List<Integer> lrpPositions) throws OpenLRProcessingException {
+        List<LocRefPoint> checkedList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/postprocessor/LrpProcessor.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package openlr.encoder.postprocessor;
+
+import openlr.OpenLRProcessingException;
+import openlr.encoder.data.LocRefPoint;
+
+import java.util.List;
+
+public interface LrpProcessor {
+    /**
+     * This method add intermediate points between the adjacent location reference points
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/LrpProcessor.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 3 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 4 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 5 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/test/java/openlr/encoder/postprocessor/AlternatePathLrpProcessorTest.java (hunk 6 above)
Comment (BonyBabu) [reply encoder/src/test/java/openlr/encoder/postprocessor/AlternatePathLrpProcessorTest.java]: d14a5b816e9b91dd99d4db70246353f967710764

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: 9356b1167ff65c403db3a957e125e48f5e976a67

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Generate a linked list of location reference points where No Partial or fully joined  alternate path of length under the given threshold
+     * exist between the road segments of adjacent location reference points
+     * @param lrps linked list of estimated location reference points
+     * @return linked list of revised lrp point linked list with intermediate points
+     * @throws OpenLRProcessingException
+     */
+    public List<LocRefPoint> process(List<LocRefPoint> lrps) throws OpenLRProcessingException {
+
+        if (properties.insertLrpAtAlternatePath()) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: To make the code better readable I would revert the logic:
- Instead of ask properties.insertLrpAtAlternatePath() and go into the logic you can better ask the if the property is not present and return, the continue the logic:

`if( !properties.insertLrpAtAlternatePath()){ return lrps; }`

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            List<LocRefPoint> revisedLrpList = new ArrayList<>();
+            for (int index = 0; index < lrps.size() - 1; ++index) {
+                LocRefPoint lrp = lrps.get(index);
+                Line firstLineOfNextLrp = lrps.get(index + 1).getLine();
+                List<Line> oldRoute = lrp.getRoute();
+                List<Line> connectedRoute = new ArrayList<>();
+                for (Line line : oldRoute) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: you can replace this for loop with:

` List<Line> connectedRoute = new ArrayList<>(oldRoute);`

then remove line 82 until 85

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+import openlr.encoder.properties.OpenLREncoderProperties;
+import openlr.encoder.routesearch.SecondShortestRouteChecker;
+import openlr.map.Line;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <h1>Processor to insert intermediate points along the shortest route between adjacent location reference points
+ * where there are alternate paths with length under the threshold exist</h1>
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: please add @since

File: encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperties.java [hunk 16]
```diff
@@ -81,6 +81,10 @@
     /** The comp time4 cache. */
     private final int compTime4Cache;
 
+    private boolean insertLrpAtAlternatePath;
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperties.java]: please add short java docs explanation for both new properties

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+import openlr.map.Line;
+import openlr.map.utils.GeometryUtils;
+import openlr.map.utils.PQElem;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * <h1>Verifying the existence of Fully or partially joined second shortest route with length less than relative threshold</h1>
+ */
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: please add `@since`

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+public class SecondShortestRouteChecker {
+    private GeoCoordinates destination;
+    private List<? extends Line> location;
+    private int maxLengthAllowed;
+    private double relativeThreshold;
+
+
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: please remove overall unnecessary empty line. Let only one and not two

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     *
+     * @param lengthAlongSecondShortestRoute length of the second shortest route
+     * @param lengthAlongLocation length of the route along location to which the second shortest route gives alternative to
+     * @return true: if the length of the second shortest route is not greater than route along location by relative threshold percentage
+     *         false: if the length of the second shortest route is greater than route along location by relative threshold percentage
+     */
+    boolean VerifyThreshold(int lengthAlongSecondShortestRoute, int lengthAlongLocation){
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: you can add the `private` accessor here

Comment (stephencurran) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: Also method names should begin with a lower case letter.

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java (hunk 13 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: fixed in 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java (hunk 14 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java (hunk 15 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AlternatePathLrpProcessor.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperties.java (hunk 16 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/properties/OpenLREncoderProperties.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 17 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 18 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 19 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: 00c70cf1bd4f3e32f47bccb0523ebe37cbbce6e8

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     *
+     * @param index index of the road segment in the location between the two location reference points
+     * @return true if an alternate path of length under the the relative threshold exist
+     *         false if no alternate path of length under the the relative threshold exist
+     */
+    public boolean hasValidDeviationBefore(int index) {
+        if (index > 0 && index < location.size() - 1) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: you can replace this if..else with
`  if (index <= 0 ){ return false;}`
then the calculation logic.

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+
+    /**
+     *
+     * @param current parent road segment
+     * @param closedSet Set of road segments to skip
+     * @return list of successor road segments
+     */
+    private List<Line> getAcceptableSuccessors(PQElem current, Set<Long> closedSet) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: I would rename the method to: ** calculateAcceptableSuccessors** 
Rename **current** to **currentRoadElement**
Rename ** closedSet** to ** closedRoadSegmentSet**

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+     * @param parent parent road segment
+     */
+    private void updateRouteSearchData(RouteSearchData data, List<Line> children, PQElem parent) {
+        for (Line child : children) {
+
+            int distanceCoveredByChild = parent.getSecondVal() + child.getLineLength();
+            int heuristics = distanceCoveredByChild + (int) GeometryUtils.distance(child.getEndNode().getGeoCoordinates(),
+                    destination);
+
+            if (data.hasLengthValue(child)) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: maybe you can invert the condition and simplify the code as follow:

if (!data.hasLengthValue(child)) {
	data.addToOpen(new PQElem(child, heuristics, distanceCoveredByChild, parent));
} else if (data.getLengthValue(child) < distanceCoveredByChild) {
	PQElem newElem = new PQElem(child, heuristics, distanceCoveredByChild, parent);
	data.updateInOpen(newElem);
}

make sure if the corner cases are working with enough test coverage ;-)

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * length of the alternate path
+     * @param destination road segment where the alternate path join back to the actual location
+     * @param index index of the road segment where alternate path starts
+     * @return length of the alternate bathe between start and route segments of alternate path.
+     */
+    private int lengthOfSecondShortestRoute(final PQElem destination, final int index) {
+        Long deviationStart = location.get(index - 1).getID();
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: use primitive -> long

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+        }
+    }
+
+    /**
+     * length of the alternate path
+     * @param destination road segment where the alternate path join back to the actual location
+     * @param index index of the road segment where alternate path starts
+     * @return length of the alternate bathe between start and route segments of alternate path.
+     */
+    private int lengthOfSecondShortestRoute(final PQElem destination, final int index) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: you can rename this method to **calculateLengthOfSecondShortestRoute**

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+                int subLocationLength = location.subList(index, destinationIndexOnLocation + index).stream().mapToInt(Line::getLineLength).sum();
+                int secondShortestRouteLength = lengthOfSecondShortestRoute(parent, index);
+                return VerifyThreshold(secondShortestRouteLength, subLocationLength);
+            } else {
+                List<Line> children = getAcceptableSuccessors(parent, closedSet);
+                updateRouteSearchData(routeSearchData, children, parent);
+                closedSet.add(parent.getLine().getID());
+            }
+        }
+        return false;
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: you are returning a boolean value, the the name of you method should indicate it. You can rename for example to: **IsNetworkForSecondShortestRouteUnderThreshold**

File: encoder/src/main/java/openlr/encoder/worker/AbstractEncoder.java [hunk 26]
```diff
@@ -139,9 +140,12 @@ public abstract RawLocationReference doEncoding(final Location location,
 
         int maxLength = properties.getMaximumDistanceLRP();
 
+        AlternatePathLrpProcessor alternatePathProcessor = AlternatePathLrpProcessor.with(properties);
+
         // find shortest-path(s) until the whole location is covered by a
         // concatenation of these shortest-path(s)
         while (!remainingLocation.isEmpty()) {
+
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/worker/AbstractEncoder.java]: please remove unnecessary empty line

File: encoder/src/test/java/openlr/encoder/routesearch/SecondShortestRouteCheckerTest.java [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+    private OpenLRMapDatabaseAdaptor map;
+
+
+   @BeforeTest
+   public void loadMapStub(){
+       InputStream mapFile = OpenLRMapDatabaseAdaptor.class.getClassLoader().getResourceAsStream("teststubs/TestMapStub.xml");
+       this.map = OpenLRMapDatabaseAdaptor.from(mapFile);
+   }
+
```
Comment (cbaeza) [inline encoder/src/test/java/openlr/encoder/routesearch/SecondShortestRouteCheckerTest.java]: please format you code correctly using the code formatter and remove all unnecessary empty lines.

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 20 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 21 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 22 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 23 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 24 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java (hunk 25 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

File: encoder/src/test/java/openlr/encoder/routesearch/SecondShortestRouteCheckerTest.java (hunk 27 above)
Comment (BonyBabu) [reply encoder/src/test/java/openlr/encoder/routesearch/SecondShortestRouteCheckerTest.java]: ac5407dc43f6f89b768b92e9571ed9f04942a5d2

Comment (BonyBabu) [reply encoder/src/test/java/openlr/encoder/routesearch/SecondShortestRouteCheckerTest.java]: edafae3f17e96589fda1e8368ec15ddaff1f9fe6

File: encoder/src/main/java/openlr/encoder/worker/AbstractEncoder.java (hunk 26 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/worker/AbstractEncoder.java]: ac5407d

File: encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+        return new SecondShortestRouteChecker(destinationStart, location, maxLengthAllowed, relativeThreshold);
+    }
+
+    /**
+     * @param lengthAlongSecondShortestRoute length of the second shortest route
+     * @param lengthAlongLocation            length of the route along location to which the second shortest route gives alternative to
+     * @return true: if the length of the second shortest route is not greater than route along location by relative threshold percentage
+     * false: if the length of the second shortest route is greater than route along location by relative threshold percentage
+     */
+    private boolean verifyThreshold(int lengthAlongSecondShortestRoute, int lengthAlongLocation) {
```
Comment (cbaeza) [inline encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: in general: all method returning a boolean have to start with the prefix **is** or **has**.
So you code have more sense and is easy to read/understand.

Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java]: 9f36acca3930e7a7842c6125aa352ec9a0bdc38f
